src/core/mesh/qgsmeshdataprovider.cpp

```diff
@@ -336,3 +337,8 @@ int QgsMesh::faceCount() const
 {
   return faces.size();
 }
+
+bool QgsMeshDataSourceInterface::meshHasChanged()
```

Comment:
This should be signal similar to `datasetGroupsAdded`, not a function. So when mesh is changed, provider will emit signal and Mesh layer will catch it in slot and update mesh layer accrodingly.

Reply:
First, I was thinking about inform the mesh layer by signals/slots mechanism. However, I think it is better to avoid this dependencies with the Qt framework and provide a better abstraction for the mechanism wich have to change the mesh. 

Does the mesh layer have to be informed as soon as the structured mesh has changed ? May be not. How I understand the concept, the provider have to supply  the data when the layer needs its, and it is not necessary to inform the layer as soon as the mesh has changed. 

With my PR, if the provider is still a QObject, it can handle a mechanism (the editor) wich manipulate the mesh data. This mechanism hasn't to be a QObject and can't emit signals. He could have no dependencies with the provider and have only to have a marker to inform the mesh has changed.

But maybe my approach is wrong ?

Reply:
we should use signal/slots in QGIS code, it is more readable, maintainable and it will work properly... Not sure how your `QgsMeshDataSourceInterface::meshHasChanged()` is implemented, but by look of it it is only workaround to proper signal. 

if you want to do a bunch of edits and then reload mesh, just emit the meshDataChanged after last edit in your plugin. if you want to do it after every edit, just do it that way.